pubsub/redis.go

```diff
@@ -17,34 +20,109 @@ const (
-	log              *log.Entry
+	node                      *node.Node
+	url                       string
+	sentinels                 string
+	sentinelDiscoveryInterval int
+	channel                   string
+	reconnectAttempt          int
+	log                       *log.Entry
 }
 
 // NewRedisSubscriber returns new RedisSubscriber struct
-func NewRedisSubscriber(node *node.Node, url string, channel string) RedisSubscriber {
+func NewRedisSubscriber(node *node.Node, url string, sentinels string, sentinelDiscoveryInterval int, channel string) RedisSubscriber {
 	return RedisSubscriber{
-		node:             node,
-		url:              url,
-		channel:          channel,
-		reconnectAttempt: 0,
-		log:              log.WithFields(log.Fields{"context": "pubsub"}),
+		node:                      node,
+		url:                       url,
+		sentinels:                 sentinels,
+		sentinelDiscoveryInterval: sentinelDiscoveryInterval,
+		channel:                   channel,
+		reconnectAttempt:          0,
+		log:                       log.WithFields(log.Fields{"context": "pubsub"}),
 	}
 }
 
 // Start connects to Redis and subscribes to the pubsub channel
+// if sentinels is set it gets the the master address first
 func (s *RedisSubscriber) Start() error {
-	// Check that URL is correct first
-	_, err := url.Parse(s.url)
+	// parse URL and check if it is correct
+	redisUrl, err := url.Parse(s.url)
 
 	if err != nil {
 		return err
 	}
 
+	var sntnl *sentinel.Sentinel
+
+	if s.sentinels != "" {
+		masterName := redisUrl.Hostname()
+
+		s.log.Debug("Redis sentinel enabled")
+		s.log.Debugf("Redis sentinel parameters:  sentinels: %s,  masterName: %s", s.sentinels, masterName)
+		sentinels := strings.Split(s.sentinels, ",")
+		sntnl = &sentinel.Sentinel{
+			Addrs:      sentinels,
+			MasterName: masterName,
+			Dial: func(addr string) (redis.Conn, error) {
+				timeout := 500 * time.Millisecond
+
+				c, err := redis.Dial(
+					"tcp",
+					addr,
+					redis.DialConnectTimeout(timeout),
+					redis.DialReadTimeout(timeout),
+					redis.DialReadTimeout(timeout),
+				)
+				if err != nil {
+					s.log.Debugf("Failed to connect to sentinel %s", addr)
+					return nil, err
+				}
+				s.log.Debugf("Successfully connected to sentinel %s", addr)
+				return c, nil
+			},
+		}
+
+		defer sntnl.Close()
+
+		// Periodically discover new Sentinels.
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		go func() {
+			err := sntnl.Discover()
+			if err != nil {
+				s.log.Warn("Failed to discover sentinels")
+			}
+			for {
+				select {
+				case <-ctx.Done():
+					return
+
+				case <-time.After(30 * time.Second):
+					err := sntnl.Discover()
+					if err != nil {
+						s.log.Warn("Failed to discover sentinels")
+					}
+				}
+			}
+		}()
+	}
+
 	for {
+
+		if s.sentinels != "" {
+			masterAddress, err := sntnl.MasterAddr()
+
+			if err != nil {
+				s.log.Warn("Failed to get master address from sentinel.")
+				return err
```

Comment:
Thanks for looking into this. Need to check the demo app. I've been testing this in our test environment before but since I rebased to 1.0-dev it stopped working. Guess I'd have to update also the ruby side.
I'll test this with the demo app as soon I got some spare time.

Reply:
Yeah, you need `1.0-dev` branch of `anycable` Ruby gem

Reply:
Hi. 
Unfortunately I cannot reproduce the issue with reconnection after I stop the master.
[Here](https://gist.github.com/rolandg/13eba4174f62ff0b24d1a0d190609a88) is a gist with the logs. 
I'll check the auto retry when sentinels are not yet available.

Reply:
Ah I see. You are using the pause command. 
That does trigger a failover but doesn't interrupt the connection from anycable to the master. If you use docker stop the failover works.

Reply:
Yes we can move it that would be more clean. Would you use function parameters for `redisUrl` and `sntnl` or write it to the `RedisSubscriber` object?

An alternative would be just drop the return. This works but does an unnecessary connection attempt to localhost:6379:
```
anycable_1        | W 2020-05-04T15:18:17.136Z context=pubsub Failed to get master address from sentinel.
anycable_1        | D 2020-05-04T15:18:17.136Z context=pubsub Got master address from sentinel: 
anycable_1        | W 2020-05-04T15:18:37.139Z context=pubsub Redis connection failed: dial tcp 127.0.0.1:6379: connect: connection refused
anycable_1        | I 2020-05-04T15:18:37.139Z context=pubsub Next Redis reconnect attempt in 1m21s
```